### PR TITLE
Redesign JoinLattice

### PR DESCRIPTION
**Pull request by fabianbs96**
IDE analyses need to define a JoinLattice over the edge-value domain. This is currently done by defining the IDE-problem as JoinLattice is a base-class of IDETabulationProblem. However, this has a couple of drawbacks:
- Conceptually implementing a JoinLattice should be a trait on the edge-value type, not on the IDEProblem
- Some optimizations are missed, for example AllBottom needs to explicitly store the bottom value
- Most of the composeWith and joinWith functions on EdgeFunctions cannot be properly defaulted

This PR has the following contributions:
- Makes JoinLattice a trait on the edge-value's type. This feature is opt-in and in case it is provided, the old representation is defaulted in terms of the new rep
- Optimize default edge functions (AllBottom, AllTop, EFComposer, ...)
- Implement a new default edge function: ConstantEdgeFunction. Nearly all IDE analyses need this to generate values from zero.  To ease constant function handling, EdgeFunctions now have a new virtual member function `bool isConstant()` that tells whether the output of `computeTarget` depends on the source edge value.
- The new Join Lattice representation together with the new and improved default edge functions have been implemented for the IDELinearConstantAnalysis as a proof of concept

### CODE REVIEW

File: include/phasar/PhasarLLVM/DataFlowSolver/IfdsIde/EdgeFunctionUtils.h [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+template <typename L, typename Enable>
+auto ConstantEdgeFunction<L, Enable>::joinWith(
+    EdgeFunctionPtrType OtherFunction) -> EdgeFunctionPtrType {
+  if (this == OtherFunction.get() || equal_to(OtherFunction)) {
+    return OtherFunction;
+  }
+
+  if (!OtherFunction->isConstant()) {
+    return AllBottom<L>::getInstance();
+  }
```
Comment (MMory) [inline include/phasar/PhasarLLVM/DataFlowSolver/IfdsIde/EdgeFunctionUtils.h:265]: should delegate to other EF, as discussed.

File: lib/PhasarLLVM/DataFlowSolver/IfdsIde/Problems/IDELinearConstantAnalysis.cpp [hunk 2]
```diff
@@ -170,19 +103,19 @@ static l_t executeBinOperation(const unsigned Op, l_t LVal, l_t RVal) {
...
     if (llvm::AddOverflow(Lop, Rop, Res)) {
-      return Bottom{};
+      return JoinLatticeTraits<l_t>::bottom();
     }
     return Res;
 
   case llvm::Instruction::Sub:
     if (llvm::SubOverflow(Lop, Rop, Res)) {
-      return Bottom{};
+      return JoinLatticeTraits<l_t>::bottom();
```
Comment (MMory) [inline lib/PhasarLLVM/DataFlowSolver/IfdsIde/Problems/IDELinearConstantAnalysis.cpp:112]: same

File: lib/PhasarLLVM/DataFlowSolver/IfdsIde/Problems/IDELinearConstantAnalysis.cpp [hunk 3]
```diff
@@ -191,17 +124,17 @@ static l_t executeBinOperation(const unsigned Op, l_t LVal, l_t RVal) {
     if (Lop == std::numeric_limits<int64_t>::min() &&
         Rop == -1) { // Would produce and overflow, as the complement of min is
                      // not representable in a signed type.
-      return Bottom{};
+      return JoinLatticeTraits<l_t>::bottom();
     }
     if (Rop == 0) { // Division by zero is UB, so we return Bot
-      return Bottom{};
+      return JoinLatticeTraits<l_t>::bottom();
```
Comment (MMory) [inline lib/PhasarLLVM/DataFlowSolver/IfdsIde/Problems/IDELinearConstantAnalysis.cpp:130]: same

File: lib/PhasarLLVM/DataFlowSolver/IfdsIde/Problems/IDELinearConstantAnalysis.cpp [hunk 4]
```diff
@@ -191,17 +124,17 @@ static l_t executeBinOperation(const unsigned Op, l_t LVal, l_t RVal) {
...
     }
     return Lop / Rop;
 
   case llvm::Instruction::URem:
   case llvm::Instruction::SRem:
     if (Rop == 0) { // Division by zero is UB, so we return Bot
-      return Bottom{};
+      return JoinLatticeTraits<l_t>::bottom();
```
Comment (MMory) [inline lib/PhasarLLVM/DataFlowSolver/IfdsIde/Problems/IDELinearConstantAnalysis.cpp:137]: same

File: include/phasar/PhasarLLVM/DataFlowSolver/IfdsIde/EdgeFunctionUtils.h [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+ * Contributors:
+ *     Philipp Schubert, Fabian Schiebel and others
+ *****************************************************************************/
+
+#ifndef PHASAR_PHASARLLVM_DATAFLOWSOLVER_IFDSIDE_EDGEFUNCTIONUTILS_H_
+#define PHASAR_PHASARLLVM_DATAFLOWSOLVER_IFDSIDE_EDGEFUNCTIONUTILS_H_
+
+#include "phasar/PhasarLLVM/DataFlowSolver/IfdsIde/EdgeFunctions.h"
+#include "phasar/PhasarLLVM/DataFlowSolver/IfdsIde/JoinLattice.h"
+#include "phasar/PhasarLLVM/Utils/ByRef.h"
```
Comment (vulder) [inline include/phasar/PhasarLLVM/DataFlowSolver/IfdsIde/EdgeFunctionUtils.h:15]:
```suggestion
#include "phasar/PhasarLLVM/Utils/ByRef.h"

```

File: include/phasar/PhasarLLVM/DataFlowSolver/IfdsIde/EdgeFunctionUtils.h [hunk 6]
```diff
@@ -0,0 +1,320 @@
...
+
+  auto OtherVal = OtherFunction->computeTarget(JoinLatticeTraits<L>::top());
+  auto JoinedVal = JoinLatticeTraits<L>::join(Value, OtherVal);
+
+  if (JoinLatticeTraits<L>::bottom() == JoinedVal) {
+    return AllBottom<L>::getInstance();
+  }
+
+  if (JoinLatticeTraits<L>::top() == JoinedVal) [[unlikely]] {
+    // XXX: Can this ever happen?
```
Comment (vulder) [inline include/phasar/PhasarLLVM/DataFlowSolver/IfdsIde/EdgeFunctionUtils.h:275]: If your not sure about the case, maybe put some assert here, to verify the semantics should it be triggered in a case.

File: include/phasar/PhasarLLVM/DataFlowSolver/IfdsIde/EdgeFunctionUtils.h [hunk 7]
```diff
@@ -0,0 +1,320 @@
...
+  auto NextVal = SecondFunction->computeTarget(Value);
+  if (NextVal == Value) {
+    return this->shared_from_this();
+  }
+
+  if (NextVal == JoinLatticeTraits<L>::bottom()) {
+    return AllBottom<L>::getInstance();
+  }
+  if (NextVal == JoinLatticeTraits<L>::top()) [[unlikely]] {
+    // XXX: Can this ever happen?
```
Comment (vulder) [inline include/phasar/PhasarLLVM/DataFlowSolver/IfdsIde/EdgeFunctionUtils.h:310]: same

File: include/phasar/PhasarLLVM/Utils/BinaryDomain.h [hunk 8]
```diff
@@ -17,6 +17,7 @@
 #ifndef PHASAR_PHASARLLVM_UTILS_BINARYDOMAIN_H_
 #define PHASAR_PHASARLLVM_UTILS_BINARYDOMAIN_H_
 
+#include "phasar/PhasarLLVM/DataFlowSolver/IfdsIde/JoinLattice.h"
```
Comment (vulder) [inline include/phasar/PhasarLLVM/Utils/BinaryDomain.h:20]:
```suggestion
#include "phasar/PhasarLLVM/DataFlowSolver/IfdsIde/JoinLattice.h"

```

**Review by vulder — APPROVED** [review]

**Comment by MMory** [comment]
This is a breaking change, isn't it? Before merging BreakingChanges.md should be updated then.

**Review by MMory — APPROVED** [review]
lgtm

**Comment by MMory** [comment]
This is a breaking change, isn't it? Before merging BreakingChanges.md should be updated then.

**Comment by fabianbs96** [comment]
This should be completely opt-in

**Pull request merged**